### PR TITLE
MGMT-7153 Support multiplatform test and starting vsphere platform basic flow

### PR DESCRIPTION
**Pull request by michaellevy101**

### CODE REVIEW

**Comment by eliorerz** [comment]
@michaellevy101 Add description on the PR

File: discovery-infra/test_infra/controllers/node_controllers/libvirt_controller.py [hunk 1]
```diff
@@ -22,9 +24,10 @@
 class LibvirtController(NodeController, ABC):
     TEST_DISKS_PREFIX = "ua-TestInfraDisk"
 
-    def __init__(self, private_ssh_key_path: Path):
+    def __init__(self, config: ControllerConfig, cluster_config: BaseClusterConfig):
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/libvirt_controller.py:27]: You changed the signature but some of usages left unchanged  e.g. `download_logs.py`

File: discovery-infra/test_infra/helper_classes/config/controller_config.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+
+@dataclass
+class ControllerConfig(_BaseConfig, ABC):
+    platform: str = None
+    base_dns_domain: str = None   # Todo - Might change during MGMT-5370
+    is_ipv6: bool = None   # Todo - Might change during MGMT-5370
+    bootstrap_in_place: bool = None
+    private_ssh_key_path: Path = None
+    working_dir: str = consts.WORKING_DIR
```
Comment (eliorerz) [inline discovery-infra/test_infra/helper_classes/config/controller_config.py:16]: Set to `None` and get set the value on `global_variables` object

File: discovery-infra/test_infra/helper_classes/config/controller_config.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+from abc import ABC
+from pathlib import Path
+from dataclasses import dataclass
+
+from test_infra import consts
+from .base_config import _BaseConfig
+
+
+@dataclass
+class ControllerConfig(_BaseConfig, ABC):
```
Comment (eliorerz) [inline discovery-infra/test_infra/helper_classes/config/controller_config.py:10]: Rename to `BaseControllerConfig` to keep with the conventions

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            return Node(name=terraform_vm_state["name"], private_ssh_key_path=self._config.private_ssh_key_path,
+                        node_controller=self)
+
+        return list(map(vsphere_vm_to_node, vms))
+
+    def destroy_all_nodes(self) -> None:
+        self._tf.destroy()
+
+    def shutdown_node(self, node_name: str) -> None:
+        pass
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:46]: Raise `NotImplementedError`

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    def shutdown_all_nodes(self) -> None:
+        pass
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:49]: Raise `NotImplementedError`

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    def start_node(self, node_name: str, check_ips: bool) -> None:
+        pass
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:52]: Same here, and for all unimplemented abstract methods

File: discovery-infra/test_infra/helper_classes/cluster.py [hunk 7]
```diff
@@ -660,8 +660,7 @@ def _set_dns(self, api_vip, ingress_vip):
     @JunitTestCase()
     def prepare_for_installation(self, static_network_config=None, **kwargs):
         self.update_config(**kwargs)
-        log.info(f"Preparing for installation with clsuter configurations: cluster_config={self._config}")
-        log.info(f"Preparing for installation with nodes configurations: nodes_config={self.nodes.config}")
```
Comment (eliorerz) [inline discovery-infra/test_infra/helper_classes/cluster.py:664]: Why not logging it?

File: discovery-infra/test_infra/helper_classes/nodes.py [hunk 8]
```diff
@@ -21,16 +21,11 @@ def __init__(self, node, cluster_host):
 class Nodes:
     DEFAULT_STATIC_IP_CONFIG = False
 
-    def __init__(self, node_controller: NodeController, private_ssh_key_path: Path):
+    def __init__(self, node_controller: NodeController):
```
Comment (eliorerz) [inline discovery-infra/test_infra/helper_classes/nodes.py:24]: This one breaks `test_scaleup` 
FYI @lalon4 @nshidlin

File: discovery-infra/test_infra/tools/terraform_utils.py [hunk 9]
```diff
@@ -20,13 +23,32 @@ def __init__(self, working_dir: str):
     def init_tf(self) -> None:
         self.tf.cmd("init -plugin-dir=/root/.terraform.d/plugins/", raise_on_error=True)
 
+    def select_defined_variables(self, **kwargs):
```
Comment (eliorerz) [inline discovery-infra/test_infra/tools/terraform_utils.py:26]: Document what `kwargs` are in this function

File: discovery-infra/test_infra/tools/terraform_utils.py [hunk 10]
```diff
@@ -20,13 +23,32 @@ def __init__(self, working_dir: str):
...
+        return results
+
     def apply(self, refresh: bool = True) -> None:
         return_value, output, err = self.tf.apply(no_color=IsFlagged, refresh=refresh, input=False, skip_plan=True)
         if return_value != 0:
             message = f"Terraform apply failed with return value {return_value}, output {output} , error {err}"
             logging.error(message)
             raise Exception(message)
 
+    def set_and_apply(self, refresh: bool = True, **kwargs) -> None:
```
Comment (eliorerz) [inline discovery-infra/test_infra/tools/terraform_utils.py:47]: Document what `kwargs` are in this function

File: discovery-infra/test_infra/utils/terraform_util.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+from test_infra.consts import consts
+from test_infra.utils import utils
+
+
+class TerraformControllerUtil:
+
+    @classmethod
+    def get_folder(cls, cluster_name: str, namespace=None):
+        folder_name = f'{cluster_name}__{namespace}' if namespace else f'{cluster_name}'
```
Comment (eliorerz) [inline discovery-infra/test_infra/utils/terraform_util.py:12]: Double quotes

File: discovery-infra/tests/base_test.py [hunk 12]
```diff
@@ -34,29 +38,105 @@
...
+    @pytest.fixture
+    def controller(self, cluster_configuration, prepare_controller_configuration):
+        if cluster_configuration.platform == consts.Platforms.VSPHERE:
+            return VSphereController(prepare_controller_configuration, cluster_configuration)
+
+        return TerraformController(prepare_controller_configuration, cluster_config=cluster_configuration)
+
+    @pytest.fixture
+    def nodes(self, controller):
+        return Nodes(controller)
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:73]: I don't think we need it in separate fixture

File: discovery-infra/tests/base_test.py [hunk 13]
```diff
@@ -34,29 +38,105 @@
...
+
+    @pytest.fixture
+    def controller_configuration(self):
+        if global_variables.platform == consts.Platforms.VSPHERE:
+            return VSphereControllerConfig()
+
+        return TerraformConfig()
+
+    @pytest.fixture
+    def prepare_controller_configuration(self, controller_configuration):
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:50]: Add type hint and return hint annotations

File: discovery-infra/tests/base_test.py [hunk 14]
```diff
@@ -34,29 +38,105 @@
...
+            return VSphereController(prepare_controller_configuration, cluster_configuration)
+
+        return TerraformController(prepare_controller_configuration, cluster_config=cluster_configuration)
+
+    @pytest.fixture
+    def nodes(self, controller):
+        return Nodes(controller)
+
+    @pytest.fixture
+    def prepare_nodes(self, nodes, cluster_configuration):
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:76]: Add type hint and return hint annotations

File: discovery-infra/tests/base_test.py [hunk 15]
```diff
@@ -34,29 +38,105 @@
...
+        tf_config.net_asset = net_asset.get()
+        yield tf_config
+        net_asset.release_all()
+
+    @pytest.fixture
+    def cluster_configuration(self):
+        return ClusterConfig()
+
+    @pytest.fixture
+    def controller(self, cluster_configuration, prepare_controller_configuration):
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:65]: Add type hint and return hint annotations

File: discovery-infra/tests/base_test.py [hunk 16]
```diff
@@ -34,29 +38,105 @@
...
+        if cluster_configuration.platform == consts.Platforms.VSPHERE:
+            return VSphereController(prepare_controller_configuration, cluster_configuration)
+
+        return TerraformController(prepare_controller_configuration, cluster_config=cluster_configuration)
+
+    @pytest.fixture
+    def nodes(self, controller):
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:72]: Add type hint and return hint annotations

File: discovery-infra/tests/base_test.py [hunk 17]
```diff
@@ -34,29 +38,105 @@
...
+        nodes.prepare_nodes()
+        yield nodes
+        if global_variables.test_teardown:
+            logging.info('--- TEARDOWN --- node controller\n')
+            nodes.destroy_all_nodes()
+            logging.info(f'--- TEARDOWN --- deleting iso file from: {cluster_configuration.iso_download_path}\n')
+            infra_utils.run_command(f"rm -f {cluster_configuration.iso_download_path}", shell=True)
+
+    @pytest.fixture
+    def prepare_network(self, prepare_nodes, prepare_controller_configuration):
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:86]: Add type hint and return hint annotations

File: discovery-infra/tests/test_e2e_install.py [hunk 18]
```diff
@@ -1,32 +1,40 @@
...
 from tests.base_test import BaseTest
-from tests.config import ClusterConfig, TerraformConfig
+from tests.config import ClusterConfig
 from tests.conftest import get_available_openshift_versions, get_api_client
 
 
 class TestInstall(BaseTest):
 
+    @pytest.fixture
+    def cluster_configuration(self, request):
```
Comment (eliorerz) [inline discovery-infra/tests/test_e2e_install.py:17]: Why? 
Just set it on the test itself. One line vs ten...

**Review by eliorerz — COMMENTED** [review]
I think this implementation (`BaseTest` implementation) give us more building blocks (fixtures to use) but add more and unneeded complexity.
Also, the `get_nodes` flow was left unchanged so it is basically duplicate flow of the current one.

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    def start_all_nodes(self) -> List[Node]:
+        self._tf.set_and_apply(**self._config.get_all(), **self._cluster_config.get_all(),
+                               cluster_name=self.cluster_name)
+        return self.list_nodes()
+
+    def list_nodes(self) -> List[Node]:
+        vms_object_type = self._tf.get_resources(resource_type="vsphere_virtual_machine")
+
+        if vms_object_type:
+            return list()
```
Comment (osherdp) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:32]:
```suggestion
            return []
```

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py (hunk 4 above)
Comment (osherdp) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:46]:
```suggestion
        raise NotImplementedError
```
and the same for all other non implemented methods

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+
+        vms = vms_object_type[0].instances
+
+        def vsphere_vm_to_node(terraform_vm_state):
+            return Node(name=terraform_vm_state["name"], private_ssh_key_path=self._config.private_ssh_key_path,
+                        node_controller=self)
+
+        return list(map(vsphere_vm_to_node, vms))
```
Comment (osherdp) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:40]:
```suggestion
        return [vsphere_vm_to_node(vm) for vm in vms]
```

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+        vms_object_type = self._tf.get_resources(resource_type="vsphere_virtual_machine")
+
+        if vms_object_type:
+            return list()
+
+        vms = vms_object_type[0].instances
+
+        def vsphere_vm_to_node(terraform_vm_state):
+            return Node(name=terraform_vm_state["name"], private_ssh_key_path=self._config.private_ssh_key_path,
+                        node_controller=self)
```
Comment (osherdp) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:38]: I usually prefer those to be private ("helper") methods of the class, but that's not a blocker imho

File: terraform_files/vsphere/main.tf [hunk 22]
```diff
@@ -0,0 +1,93 @@
+provider "vsphere" {
+  user                 = var.vsphere_username
```
Comment (YuviGold) [inline terraform_files/vsphere/main.tf:2]: Elaborate on the description of each data section.
In addition, I'd be glad if you could add a README providing an overview about using `vsphere` provider.

File: terraform_files/vsphere/main.tf [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+  type          = "vm"
+  datacenter_id = data.vsphere_datacenter.datacenter.id
+  tags          = [vsphere_tag.tag.id]
+}
+
+resource "vsphere_file" "ISO_UPLOAD" {
+  datacenter       = var.vsphere_datacenter
+  datastore        = var.vsphere_datastore
+  source_file      = var.iso_download_path
+  destination_file = "ISOs/${basename(var.iso_download_path)}"
```
Comment (YuviGold) [inline terraform_files/vsphere/main.tf:57]: The `iso_download_path` should be absolute

File: discovery-infra/test_infra/helper_classes/config/controller_config.py [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+    master_disk_count: int = None   # number of disks to create
+
+    worker_memory: int = None
+    worker_vcpu: int = None
+    workers_count: int = None
+    worker_cpu_mode: str = None
+    worker_disk: int = None
+    worker_disk_count: int = None
+
+    network_mtu: int = None
```
Comment (osherdp) [inline discovery-infra/test_infra/helper_classes/config/controller_config.py:33]: configure your IDE to add new lines on Python modules :)

File: terraform_files/vsphere/main.tf [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+  resource_pool_id     = data.vsphere_compute_cluster.cluster.resource_pool_id
+  datastore_id         = data.vsphere_datastore.datastore.id
+  num_cpus             = var.master_vcpu
+  num_cores_per_socket = var.vsphere_control_plane_cores_per_socket
+  memory               = var.master_memory
+  guest_id             = "coreos64Guest"
+  folder               = var.cluster_name
+  enable_disk_uuid     = "true"
+  wait_for_guest_net_timeout  = "0"
+  wait_for_guest_net_routable = "false"
```
Comment (YuviGold) [inline terraform_files/vsphere/main.tf:73]: Why not?

File: terraform_files/vsphere/main.tf [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+
+  network_interface {
+    network_id = data.vsphere_network.network.id
+  }
+  
+  disk {
+    label            = "disk0"
+    size             = var.master_disk_size_gib
+    eagerly_scrub    = false
+    thin_provisioned = true
```
Comment (YuviGold) [inline terraform_files/vsphere/main.tf:83]: What does it mean? please add a description.

File: discovery-infra/test_infra/helper_classes/config/controller_config.py [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+    bootstrap_in_place: bool = None
+    private_ssh_key_path: Path = None
+    working_dir: str = consts.WORKING_DIR
+
+    master_memory: int = None
+    master_vcpu: int = None
+    masters_count: int = None
+    master_cpu_mode: str = None
+    master_disk: int = None   # disk size in MB.
+    master_disk_size_gib: str = None # disk size in GB.
```
Comment (osherdp) [inline discovery-infra/test_infra/helper_classes/config/controller_config.py:23]: it doesn't sound fun to have two sources of truth
maybe save the int one, and have a small function that returns the other value?

File: terraform_files/vsphere/main.tf [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+  folder               = var.cluster_name
+  enable_disk_uuid     = "true"
+  wait_for_guest_net_timeout  = "0"
+  wait_for_guest_net_routable = "false"
+
+  network_interface {
+    network_id = data.vsphere_network.network.id
+  }
+  
+  disk {
```
Comment (YuviGold) [inline terraform_files/vsphere/main.tf:79]: Should follow naming like we have for libvirt
https://github.com/openshift/assisted-test-infra/blob/51d5d1c28822dba48ac56dd4210117d0b606a272/terraform_files/baremetal/main.tf#L100-L107

File: discovery-infra/tests/test_e2e_install.py [hunk 29]
```diff
@@ -1,32 +1,40 @@
...
 
     @JunitTestSuite()
     @pytest.mark.parametrize("sleep_time", [1, 60])
-    def test_dummy(self, configs: Tuple[ClusterConfig, TerraformConfig], get_nodes, get_cluster, sleep_time):
-        cluster_config, tf_config = configs
-        new_cluster = get_cluster(get_nodes(tf_config, cluster_config), cluster_config)
-        new_cluster.prepare_for_installation()
+    def test_dummy(self, cluster, sleep_time):
+        cluster.prepare_for_installation()
         time.sleep(sleep_time)
```
Comment (YuviGold) [inline discovery-infra/tests/test_e2e_install.py:38]: What's this test? @eliorerz :)

File: discovery-infra/test_infra/helper_classes/nodes.py (hunk 8 above)
Comment (osherdp) [reply discovery-infra/test_infra/helper_classes/nodes.py:24]: it also breaks our gathering of sos-report:
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_assisted-test-infra/949/pull-ci-openshift-assisted-test-infra-master-e2e-metal-assisted/1414354554329239552/artifacts/e2e-metal-assisted/baremetalds-assisted-gather/build-log.txt

File: discovery-infra/tests/test_e2e_install.py (hunk 18 above)
Comment (YuviGold) [inline discovery-infra/tests/test_e2e_install.py:17]: Who calls it?

File: discovery-infra/test_infra/tools/terraform_utils.py [hunk 30]
```diff
@@ -20,13 +23,32 @@ def __init__(self, working_dir: str):
     def init_tf(self) -> None:
         self.tf.cmd("init -plugin-dir=/root/.terraform.d/plugins/", raise_on_error=True)
 
+    def select_defined_variables(self, **kwargs):
+        supported_variables = self.get_variable_list()
+        return {k: v for k, v in kwargs.items() if k in supported_variables}
+
+    def get_variable_list(self):
```
Comment (osherdp) [inline discovery-infra/test_infra/tools/terraform_utils.py:30]:
```suggestion
    def get_variables(self):
```

File: discovery-infra/test_infra/controllers/node_controllers/node_controller.py [hunk 31]
```diff
@@ -3,12 +3,18 @@
...
+from test_infra.helper_classes.config.controller_config import ControllerConfig
 from test_infra.controllers.node_controllers.disk import Disk
 from test_infra.controllers.node_controllers.node import Node
 
 
 class NodeController(ABC):
 
+    def __init__(self, config: ControllerConfig, cluster_config: BaseClusterConfig):
+        self._config = config
+        self._cluster_config = cluster_config
```
Comment (YuviGold) [inline discovery-infra/test_infra/controllers/node_controllers/node_controller.py:16]: Why do the nodes needs the cluster config?

File: discovery-infra/test_infra/helper_classes/config/controller_config.py (hunk 3 above)
Comment (YuviGold) [reply discovery-infra/test_infra/helper_classes/config/controller_config.py:10]: The name should be nodes config, no? This is not just a generic config that would be used from any future controller.

File: discovery-infra/test_infra/helper_classes/config/nodes_config.py [hunk 32]
```diff
@@ -37,14 +22,10 @@ class BaseTerraformConfig(_BaseConfig, ABC):
     libvirt_worker_ips: List[str] = None
```
Comment (YuviGold) [inline discovery-infra/test_infra/helper_classes/config/nodes_config.py:22]: so this variable of `libvirt_worker_ips` shouldn't be under `TFConfig` or the name should be `LibvirtConfig`

File: discovery-infra/test_infra/tools/terraform_utils.py [hunk 33]
```diff
@@ -36,10 +58,14 @@ def change_variables(self, variables: Dict[str, str], refresh: bool = True) -> N
...
         self.apply(refresh=refresh)
 
-    def get_state(self) -> str:
+    def get_state(self) -> Tfstate:
         self.tf.read_state_file(self.STATE_FILE)
         return self.tf.tfstate
 
+    def get_resources(self, resource_type: str = None) -> list:
+        state = self.get_state()
+        return [resource for resource in state.resource if resource_type is None or resource["type"] == resource_type]
```
Comment (osherdp) [inline discovery-infra/test_infra/tools/terraform_utils.py:67]:
```suggestion
        return [resource for resource in state.resource
                if resource_type is None or resource["type"] == resource_type]
```
just a nit

File: discovery-infra/test_infra/utils/terraform_util.py [hunk 34]
```diff
@@ -0,0 +1,26 @@
+import logging
+import os
+
+from test_infra.consts import consts
+from test_infra.utils import utils
+
+
+class TerraformControllerUtil:
```
Comment (osherdp) [inline discovery-infra/test_infra/utils/terraform_util.py:8]: maybe just delete this class and put all those methods in the module?
the module itself is named ``terraform_util``

File: discovery-infra/test_infra/utils/terraform_util.py [hunk 35]
```diff
@@ -0,0 +1,26 @@
+import logging
```
Comment (osherdp) [inline discovery-infra/test_infra/utils/terraform_util.py:1]: rename module to ``terraform_utils.py``

**Review by YuviGold — COMMENTED** [review]
Semi review

Please split this PR into two PRs
* Generic infra changes
* VSphere controller addition

In addition, please provide a very detailed description of the PRs. It cannot be reviewed otherwise.

File: terraform_files/vsphere/terraform.tfvars.json [hunk 36]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (osherdp) [inline terraform_files/vsphere/terraform.tfvars.json:1]: this is an example file?
if so I would add more examples for values that usually we don't know how they should look (usually addresses/server urls that can be constructed in many ways)

File: discovery-infra/tests/base_test.py [hunk 37]
```diff
@@ -34,29 +38,105 @@
...
+        if global_variables.platform != consts.Platforms.BARE_METAL and global_variables.platform != consts.Platforms.NONE:
```
Comment (osherdp) [inline discovery-infra/tests/base_test.py:87]:
```suggestion
        if global_variables.platform not in (consts.Platforms.BARE_METAL, consts.Platforms.NONE):
```

File: discovery-infra/tests/base_test.py [hunk 38]
```diff
@@ -34,29 +38,105 @@
...
+        if request.node.result_call.failed:
+            logging.info(f'--- TEARDOWN --- Collecting Logs for test: {request.node.name}\n')
+            self.collect_test_logs(cluster, api_client, request.node, cluster.nodes)
+            if global_variables.test_teardown:
+                if cluster.is_installing() or cluster.is_finalizing():
+                    cluster.cancel_install()
+                with suppress(ApiException):
+                    logging.info(f'--- TEARDOWN --- deleting created cluster {cluster.id}\n')
+                    cluster.delete()
+
```
Comment (osherdp) [inline discovery-infra/tests/base_test.py:113]: this part needs a bit more spacing to help readability

File: terraform_files/vsphere/main.tf [hunk 39]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+data "vsphere_network" "network" {
+  name          = var.vsphere_network
+  datacenter_id = data.vsphere_datacenter.datacenter.id
+}
+
+resource "vsphere_tag_category" "category" {
+  name        = "openshift-${var.cluster_name}"
+  description = "Added by openshift-install do not remove"
```
Comment (osherdp) [inline terraform_files/vsphere/main.tf:29]: maybe expand description, or add comments to explain what's this value

File: terraform_files/vsphere/main.tf [hunk 40]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+resource "vsphere_tag" "tag" {
+  name        = var.cluster_name
+  category_id = vsphere_tag_category.category.id
+  description = "Added by openshift-install do not remove"
+}
+
+resource "vsphere_folder" "folder" {
+  count = 1
```
Comment (osherdp) [inline terraform_files/vsphere/main.tf:46]:
```suggestion
  count         = 1
```

File: terraform_files/vsphere/main.tf [hunk 41]
```diff
@@ -0,0 +1,93 @@
...
+  name                 = "${var.cluster_name}-master-${count.index}"
+  resource_pool_id     = data.vsphere_compute_cluster.cluster.resource_pool_id
+  datastore_id         = data.vsphere_datastore.datastore.id
+  num_cpus             = var.master_vcpu
+  num_cores_per_socket = var.vsphere_control_plane_cores_per_socket
+  memory               = var.master_memory
+  guest_id             = "coreos64Guest"
+  folder               = var.cluster_name
+  enable_disk_uuid     = "true"
+  wait_for_guest_net_timeout  = "0"
```
Comment (osherdp) [inline terraform_files/vsphere/main.tf:72]:
```suggestion
  wait_for_guest_net_timeout  = 0
```

**Review by osherdp — CHANGES_REQUESTED** [review]
well done for all the work
just a couple of fixes and we're good to go

File: discovery-infra/tests/base_test.py [hunk 42]
```diff
@@ -34,29 +38,105 @@
...
+        interfaces = BaseTest.nat_interfaces(prepare_controller_configuration)
+        nat = NatController(interfaces, NatController.get_namespace_index(interfaces[0]))
+        nat.add_nat_rules()
+        yield prepare_nodes
+        if nat:
+            nat.remove_nat_rules()
+
+    @pytest.fixture
+    def cluster(self, api_client, request, proxy_server, prepare_network, cluster_configuration):
```
Comment (nshidlin) [inline discovery-infra/tests/base_test.py:98]: While `prepare_network` is the last fixture in the chain it says nothing about what the fixture does.

File: discovery-infra/test_infra/helper_classes/config/controller_config.py (hunk 2 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/helper_classes/config/controller_config.py:16]: why? the work dir is constant not an env variable

File: discovery-infra/test_infra/helper_classes/config/controller_config.py (hunk 3 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/helper_classes/config/controller_config.py:10]: It's a common controller configurations not a node configuration

File: discovery-infra/tests/test_e2e_install.py (hunk 18 above)
Comment (michaellevy101) [reply discovery-infra/tests/test_e2e_install.py:17]: to allow you to override it...

File: terraform_files/vsphere/main.tf (hunk 23 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:57]: This property is the path on the vsphere itself

File: terraform_files/vsphere/main.tf (hunk 22 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:2]: It fully documented on the terraform provider why to document it here? I have to document the variables i have created or special behavior

File: terraform_files/vsphere/main.tf (hunk 26 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:83]: https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#eagerly_scrub
https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#thin_provisioned

File: discovery-infra/tests/test_e2e_install.py (hunk 29 above)
Comment (michaellevy101) [reply discovery-infra/tests/test_e2e_install.py:38]: irrelevant,  please open a new task for this test...

File: discovery-infra/tests/test_e2e_install.py (hunk 18 above)
Comment (michaellevy101) [reply discovery-infra/tests/test_e2e_install.py:17]: It's a fixture override,  I will document it

File: discovery-infra/test_infra/controllers/node_controllers/node_controller.py (hunk 31 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/node_controller.py:16]: It needs the cluster name and maybe more, for the extended controllers

File: discovery-infra/test_infra/helper_classes/config/nodes_config.py (hunk 32 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/helper_classes/config/nodes_config.py:22]: can't fix everything in one PR,  keep the original code ( If i remember correctly)

File: terraform_files/vsphere/main.tf (hunk 39 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:29]: The description is the tag description which should be applied on every resource, i should be "Created by the test-infra, do not delete manually" but as you understand i just copied it from the Openshift installer and not ready for a review (just asked to publish the code to provide a picture of the changes)..... I will document everything properly

File: terraform_files/vsphere/main.tf (hunk 41 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:72]: ?

File: discovery-infra/tests/base_test.py (hunk 42 above)
Comment (michaellevy101) [reply discovery-infra/tests/base_test.py:98]: prepare network is preparing the network.... in this case configuring the NAT. don't understand what do you mean by "says nothing about what the fixture does" can you please extend?

**Comment by michaellevy101** [comment]
@eliorerz @eliorerz @osherdp @YuviGold @nshidlin Thank you for your comments but this code isn't ready yet
I just asked to push the code to provide a picture of how to support multi-platform tests
I have addressed most of your comments

**Comment by osherdp** [comment]
> @eliorerz @eliorerz @osherdp @YuviGold @nshidlin Thank you for your comments but this code isn't ready yet
> I just asked to push the code to provide a picture of how to support multi-platform tests
> I have addressed most of your comments

No problem @michaellevy101 
next time please add [WIP] for the PR title, so people will know 
description also

And again, great work is being done here. Very nice :)

**Pull request closed without merging**